# PA005: High Value Customer Indentification ( Insiders )

In [1]:
import re
import os
import numpy as np
import pandas as pd
import seaborn as sns
import pickle
import s3fs
import boto3

import umap.umap_ as umap

from sklearn import cluster as c
from sklearn import metrics as m
from sklearn import mixture as mx
from sklearn import manifold as mn 
from sklearn import ensemble as en
from sklearn import decomposition as dd
from sklearn import preprocessing as pp

from plotly import express as px

from matplotlib import pyplot as plt

/Users/thiago/Thiago/Data_Science/comunidade_ds/insiders_clustering/ins_clustenv/lib/python3.10/site-packages/tqdm/auto.py:22: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


### 0.2. Load Dataset

In [2]:
# load data
df_raw = pd.read_csv('/Users/thiago/Thiago/Data_Science/comunidade_ds/insiders_clustering/data/raw/ecommerce.csv', encoding='unicode_escape' )

# drop extra column
df_raw = df_raw.drop( columns=['Unnamed: 8'], axis=1 )

## <font color='red'> 1.0. Descrição dos Dados </font>

In [3]:
df1 = df_raw.copy()

### 1.1. Rename Columns

In [4]:
cols_new = [ 'invoice_no', 'stock_code', 'description', 'quantity', 'invoice_date',
       'unit_price', 'customer_id', 'country' ]
df1.columns = cols_new

### 1.2. Data Dimensions

In [5]:
print( 'Number of rows: {}'.format( df1.shape[0] ) )
print( 'Number of cols: {}'.format( df1.shape[1] ) )

Number of rows: 541909
Number of cols: 8


### 1.3. Data Types

In [6]:
df1.dtypes

invoice_no       object
stock_code       object
description      object
quantity          int64
invoice_date     object
unit_price      float64
customer_id     float64
country          object
dtype: object

### 1.4. Check NA

In [7]:
df1.isna().sum()

invoice_no           0
stock_code           0
description       1454
quantity             0
invoice_date         0
unit_price           0
customer_id     135080
country              0
dtype: int64

### 1.5. Replace NA

In [8]:
df_missing = df1.loc[df1['customer_id'].isna(), :]
df_not_missing = df1.loc[~df1['customer_id'].isna(), :]

In [9]:
# create reference
df_backup = pd.DataFrame( df_missing['invoice_no'].drop_duplicates() )
df_backup['customer_id'] = np.arange( 19000, 19000+len( df_backup ), 1)

# merge original with reference dataframe
df1 = pd.merge( df1, df_backup, on='invoice_no', how='left' )

# coalesce
df1['customer_id'] = df1['customer_id_x'].combine_first( df1['customer_id_y'] )

# drop extra columns
df1 = df1.drop( columns=['customer_id_x', 'customer_id_y'], axis=1 )

df1.head()


invoice_no stock_code                          description  quantity  \
0     536365     85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1     536365      71053                  WHITE METAL LANTERN         6   
2     536365     84406B       CREAM CUPID HEARTS COAT HANGER         8   
3     536365     84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4     536365     84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

  invoice_date  unit_price         country  customer_id  
0    29-Nov-16        2.55  United Kingdom      17850.0  
1    29-Nov-16        3.39  United Kingdom      17850.0  
2    29-Nov-16        2.75  United Kingdom      17850.0  
3    29-Nov-16        3.39  United Kingdom      17850.0  
4    29-Nov-16        3.39  United Kingdom      17850.0

In [10]:
df1.isna().sum()

invoice_no         0
stock_code         0
description     1454
quantity           0
invoice_date       0
unit_price         0
country            0
customer_id        0
dtype: int64

### 1.6. Change Types

In [11]:
# invoice_date
df1['invoice_date'] = pd.to_datetime( df1['invoice_date'], format='%d-%b-%y' )

# customer_id
df1['customer_id'] = df1['customer_id'].astype(int)

In [12]:
df1.dtypes

invoice_no              object
stock_code              object
description             object
quantity                 int64
invoice_date    datetime64[ns]
unit_price             float64
country                 object
customer_id              int64
dtype: object

### 1.7. Descriptive Statistics

In [13]:
num_attributes = df1.select_dtypes( include=['int64', 'float64'] )
cat_attributes = df1.select_dtypes( exclude=['int64', 'float64', 'datetime64[ns]'] )

#### 1.7.1. Numerical Attributes

In [14]:
# central tendency - mean, median
ct1 = pd.DataFrame( num_attributes.apply( np.mean ) ).T
ct2 = pd.DataFrame( num_attributes.apply( np.median ) ).T

# dispersion - desvio padrão, mínimo, máximo, range, skew, kurtosis
d1 = pd.DataFrame( num_attributes.apply( np.std ) ).T
d2 = pd.DataFrame( num_attributes.apply( np.min ) ).T
d3 = pd.DataFrame( num_attributes.apply( np.max ) ).T
d4 = pd.DataFrame( num_attributes.apply( lambda x: x.max() - x.min() ) ).T
d5 = pd.DataFrame( num_attributes.apply( lambda x: x.skew() ) ).T
d6 = pd.DataFrame( num_attributes.apply( lambda x: x.kurtosis() ) ).T

# concatenate
m1 = pd.concat( [d2, d3, d4, ct1, ct2, d1, d5, d6] ).T.reset_index()
m1.columns = ['attributes', 'min', 'max', 'range', 'mean', 'mediana', 'std', 'skew', 'kurtosis']
m1


attributes       min      max      range          mean   mediana  \
0     quantity -80995.00  80995.0  161990.00      9.552250      3.00   
1   unit_price -11062.06  38970.0   50032.06      4.611114      2.08   
2  customer_id  12346.00  22709.0   10363.00  16688.840453  16249.00   

           std        skew       kurtosis  
0   218.080957   -0.264076  119769.160031  
1    96.759764  186.506972   59005.719097  
2  2911.408666    0.487449      -0.804287

#### 1.7.2. Categorical Attributes

In [15]:
cat_attributes.head()

invoice_no stock_code                          description         country
0     536365     85123A   WHITE HANGING HEART T-LIGHT HOLDER  United Kingdom
1     536365      71053                  WHITE METAL LANTERN  United Kingdom
2     536365     84406B       CREAM CUPID HEARTS COAT HANGER  United Kingdom
3     536365     84029G  KNITTED UNION FLAG HOT WATER BOTTLE  United Kingdom
4     536365     84029E       RED WOOLLY HOTTIE WHITE HEART.  United Kingdom

<b>Invoice No</b>

In [16]:
# problema: Temos o invoice com letras e numeros
#df1['invoice_no'].astype( int )

# Identificação
df_letter_invoices = df1.loc[df1['invoice_no'].apply( lambda x: bool ( re.search( '[^0-9]+', x ) ) ), :]
df_letter_invoices.head()

print( 'Total number os invoices: {}'.format( len ( df_letter_invoices ) ) ) 
print( 'Total number os negative quantity: {}'.format( len( df_letter_invoices[df_letter_invoices['quantity'] < 0 ] ) ) )

Total number os invoices: 9291
Total number os negative quantity: 9288


In [17]:
#cat_attributes['invoice_no'].astype( int )
len( cat_attributes.loc[ cat_attributes['invoice_no'].apply( lambda x: bool( re.search( '[^0-9]+', x ) ) ), 'invoice_no'].drop_duplicates() )
#cat_attributes.loc[173995, : ]

3839

<b>Stock Code</b>

In [18]:
# stock code
#at_least_one_string = cat_attributes.loc[ cat_attributes['stock_code'].apply( lambda x: bool ( re.search( '[^0-9+]', x ) ) ), 'stock_code'].head()

#print( at_least_one_string )

#
#cat_attributes.loc[cat_attributes['stock_code'].apply( lambda x: bool( re.search( '^[a-zA-Z]+$', x ) ) ), 'stock_code'].unique() 
df1.loc[cat_attributes['stock_code'].apply( lambda x: bool( re.search( '^[a-zA-Z]+$', x ) ) ), :].head()

invoice_no stock_code     description  quantity invoice_date  unit_price  \
45       536370       POST         POSTAGE         3   2016-11-29       18.00   
141     C536379          D        Discount        -1   2016-11-29       27.50   
386      536403       POST         POSTAGE         1   2016-11-29       15.00   
1123     536527       POST         POSTAGE         1   2016-11-29       18.00   
1814     536544        DOT  DOTCOM POSTAGE         1   2016-11-29      569.77   

             country  customer_id  
45            France        12583  
141   United Kingdom        14527  
386      Netherlands        12791  
1123         Germany        12662  
1814  United Kingdom        19001

In [19]:
# check stock codes only characters
df1.loc[df1['stock_code'].apply( lambda x: bool ( re.search( '^[a-zA-Z]+$', x ) ) ), 'stock_code'].unique()

# Ação:
## 1. Remove stock_code in ['POST', 'D', 'M', 'PADS', 'DOT', 'CRUK']

array(['POST', 'D', 'DOT', 'M', 'S', 'AMAZONFEE', 'm', 'DCGSSBOY',
       'DCGSSGIRL', 'PADS', 'B', 'CRUK'], dtype=object)

<b>Description</b>

In [20]:
df1.head()

# Ação: deletar description

invoice_no stock_code                          description  quantity  \
0     536365     85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1     536365      71053                  WHITE METAL LANTERN         6   
2     536365     84406B       CREAM CUPID HEARTS COAT HANGER         8   
3     536365     84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4     536365     84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

  invoice_date  unit_price         country  customer_id  
0   2016-11-29        2.55  United Kingdom        17850  
1   2016-11-29        3.39  United Kingdom        17850  
2   2016-11-29        2.75  United Kingdom        17850  
3   2016-11-29        3.39  United Kingdom        17850  
4   2016-11-29        3.39  United Kingdom        17850

<b>Country</b>

In [21]:
len( df1['country'].unique() )

38

In [22]:
df1['country'].value_counts( normalize=True)

United Kingdom          0.914320
Germany                 0.017521
France                  0.015790
EIRE                    0.015124
Spain                   0.004674
Netherlands             0.004375
Belgium                 0.003818
Switzerland             0.003694
Portugal                0.002803
Australia               0.002323
Norway                  0.002004
Italy                   0.001482
Channel Islands         0.001399
Finland                 0.001283
Cyprus                  0.001148
Sweden                  0.000853
Unspecified             0.000823
Austria                 0.000740
Denmark                 0.000718
Japan                   0.000661
Poland                  0.000629
Israel                  0.000548
USA                     0.000537
Hong Kong               0.000531
Singapore               0.000423
Iceland                 0.000336
Canada                  0.000279
Greece                  0.000269
Malta                   0.000234
United Arab Emirates    0.000125
European C

In [23]:
df1[['customer_id', 'country']].drop_duplicates().groupby( 'country' ).count().reset_index().sort_values( 'customer_id', ascending=False )

country  customer_id
36        United Kingdom         7587
14               Germany           95
13                France           90
10                  EIRE           44
31                 Spain           31
3                Belgium           25
33           Switzerland           24
27              Portugal           20
19                 Italy           15
16             Hong Kong           15
12               Finland           12
1                Austria           11
25                Norway           10
24           Netherlands            9
0              Australia            9
37           Unspecified            9
9                Denmark            9
6        Channel Islands            9
7                 Cyprus            8
32                Sweden            8
20                 Japan            8
18                Israel            7
26                Poland            6
5                 Canada            4
34                   USA            4
2                Bahrain            4
15                Greece            4
23                 Malta            2
35  United Arab Emirates            2
8         Czech Republic            1
22             Lithuania            1
11    European Community            1
21               Lebanon            1
28                   RSA            1
29          Saudi Arabia            1
30             Singapore            1
17               Iceland            1
4                 Brazil            1

## <font color='red'> 2.0. Filtragem das Variáveis </font>

In [24]:
df2 = df1.copy()

In [25]:
# ===== Numerical Attributes ======
# unit price > 0.04
df2 = df2.loc[df2['unit_price'] > 0.04, :]

# ===== Categorical Attributes ======
# stock code != [ POST, D, M, DOT, CRUK ]
df2 = df2[~df2['stock_code'].isin( ['POST', 'D', 'DOT', 'M', 'S', 'AMAZONFEE', 'm', 'DCGSSBOY',
       'DCGSSGIRL', 'PADS', 'B', 'CRUK']) ] 

# description
df2 = df2.drop( columns='description', axis=1 )

# map
df2 = df2[~df2['country'].isin( ['European Community', 'Unspecified'] ) ]

# bad users
df2 = df2[~df2['customer_id'].isin( [16446] ) ] 

## quantity - Negative numbers means product returns
df2_returns = df2.loc[df2['quantity'] < 0, :]
df2_purchase = df2.loc[df2['quantity'] >= 0, :]

## <font color='red'> 3.0. Feature Engineering </font>

In [26]:
df3 = df2.copy()

### 3.1. Feature Creation

In [27]:
# data reference
df_ref = df3.drop( ['invoice_no', 'stock_code','quantity', 'invoice_date', 'unit_price', 'country' ], axis=1 ).drop_duplicates( ignore_index=True)

#### 3.1.1. Gross Revenue

In [28]:
# Gross Revenue ( Faturamento ) quantity * price
df2_purchase.loc[:, 'gross_revenue'] = df2_purchase.loc[:, 'quantity'] * df2_purchase.loc[:, 'unit_price']

# Monetary
df_monetary = df2_purchase.loc[:, ['customer_id', 'gross_revenue']].groupby( 'customer_id').sum().reset_index()
df_ref = pd.merge( df_ref, df_monetary, how='left', on='customer_id' )
df_ref.isna().sum()

/var/folders/w5/0w2xv4r95g1gfgbzpp2dm6gr0000gn/T/ipykernel_10240/225208262.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2_purchase.loc[:, 'gross_revenue'] = df2_purchase.loc[:, 'quantity'] * df2_purchase.loc[:, 'unit_price']


customer_id       0
gross_revenue    91
dtype: int64

#### 3.1.2. Recency - Day from last Purchase

In [29]:
# Recency - It depends on product returns
df_recency = df2_purchase.loc[:, ['customer_id', 'invoice_date']].groupby( 'customer_id' ).max().reset_index()
df_recency['recency_days'] = ( df2_purchase['invoice_date'].max() - df_recency['invoice_date'] ).dt.days
df_recency = df_recency[['customer_id', 'recency_days']].copy()
df_ref = pd.merge( df_ref, df_recency, how='left', on='customer_id' )
df_ref.isna().sum()

customer_id       0
gross_revenue    91
recency_days     91
dtype: int64

#### 3.1.3. Quantity of Purchased 

In [30]:
# Número de Produtos - It depends os product returns
df_frequency = (df2_purchase.loc[:, ['customer_id', 'invoice_no']].drop_duplicates()
                                                                  .groupby( 'customer_id' )
                                                                  .count()
                                                                  .reset_index()
                                                                  .rename( columns={'invoice_no':'qtde_invoices'}))

df_ref = pd.merge( df_ref, df_frequency, how='left', on='customer_id' )
df_ref.isna().sum()

customer_id       0
gross_revenue    91
recency_days     91
qtde_invoices    91
dtype: int64

#### 3.1.4. Quantity of Items Purchased 

In [31]:
# Número de Produtos - It depends os product returns
df_frequency = (df2_purchase.loc[:, ['customer_id', 'quantity']].groupby( 'customer_id' )
                                                                .sum()
                                                                .reset_index()
                                                                .rename( columns={'quantity':'qtde_items'} ))
df_ref = pd.merge( df_ref, df_frequency, how='left', on='customer_id' )
df_ref.isna().sum()

customer_id       0
gross_revenue    91
recency_days     91
qtde_invoices    91
qtde_items       91
dtype: int64

#### 3.1.5. Quantity of Products Purchased

In [32]:
# Número de Produtos - It depends os product returns
df_frequency = (df2_purchase.loc[:, ['customer_id', 'stock_code']].groupby( 'customer_id' )
                                                                .count()
                                                                .reset_index()
                                                                .rename( columns={'stock_code':'qtde_products'} ))
df_ref = pd.merge( df_ref, df_frequency, how='left', on='customer_id' )
df_ref.isna().sum()

customer_id       0
gross_revenue    91
recency_days     91
qtde_invoices    91
qtde_items       91
qtde_products    91
dtype: int64

#### 3.1.6. Frequency Purchase

In [33]:
df_aux = ( df2_purchase[['customer_id', 'invoice_no', 'invoice_date']].drop_duplicates()
                                                    .groupby( 'customer_id' )
                                                    .agg( max_ = ( 'invoice_date', 'max' ),
                                                          min_ = ( 'invoice_date', 'min' ),
                                                         days_ = ( 'invoice_date', lambda x: ( ( x.max() - x.min() ).days ) + 1 ),
                                                          buy_ = ( 'invoice_no', 'count' ) ) ).reset_index()

# Frequency
df_aux['frequency'] = df_aux[[ 'buy_', 'days_']].apply( lambda x: x['buy_'] / x['days_'] if x['days_'] != 0 else 0, axis=1)

# Merge
df_ref = pd.merge( df_ref, df_aux[['customer_id', 'frequency']], on='customer_id', how='left' )

In [34]:
df_ref.isna().sum()

customer_id       0
gross_revenue    91
recency_days     91
qtde_invoices    91
qtde_items       91
qtde_products    91
frequency        91
dtype: int64

#### 3.1.7. Number os Returns

In [35]:
# Number of Returns
df_returns = df2_returns[['customer_id', 'quantity']].groupby( 'customer_id' ).sum().reset_index().rename( columns={'quantity':'qtde_returns'} )
df_returns['qtde_returns'] = df_returns['qtde_returns'] * -1

df_ref = pd.merge( df_ref, df_returns, how='left', on='customer_id' )
df_ref.loc[df_ref['qtde_returns'].isna(), 'qtde_returns'] = 0

df_ref.isna().sum()

customer_id       0
gross_revenue    91
recency_days     91
qtde_invoices    91
qtde_items       91
qtde_products    91
frequency        91
qtde_returns      0
dtype: int64

#### 3.1.7. Average Ticket Value

In [36]:
# Average Ticket
df_avg_ticket = df2_purchase.loc[:, ['customer_id', 'gross_revenue']].groupby( 'customer_id' ).mean().reset_index().rename( columns={'gross_revenue' : 'avg_ticket' } ) 
#df_avg_ticket['avg_ticket'] = np.round( df_avg_ticket['avg_ticket'], 2 )
df_ref = pd.merge( df_ref, df_avg_ticket, how='left', on='customer_id' )
df_ref.isna().sum()

customer_id       0
gross_revenue    91
recency_days     91
qtde_invoices    91
qtde_items       91
qtde_products    91
frequency        91
qtde_returns      0
avg_ticket       91
dtype: int64

#### 3.1.8. Average Recency Days

In [37]:
# Average Recency Days
df_aux = df2[['customer_id', 'invoice_date']].drop_duplicates().sort_values( ['customer_id', 'invoice_date'], ascending=[ True, True] )
df_aux['next_customer_id'] = df_aux['customer_id'].shift() # next customer
df_aux['previous_date'] = df_aux['invoice_date'].shift() #next invoice date

df_aux['avg_recency_days'] = df_aux.apply( lambda x: ( x['invoice_date'] - x['previous_date'] ).days if x['customer_id'] == x['next_customer_id'] else np.nan, axis=1 )

df_aux = df_aux.drop( ['invoice_date', 'next_customer_id', 'previous_date'], axis=1 ).dropna()

# Average Recency
df_avg_recency_days = df_aux.groupby( 'customer_id' ).mean().reset_index()

# Merge
df_ref = pd.merge( df_ref, df_avg_recency_days, on='customer_id', how='left' )
df_ref.isna().sum()

customer_id            0
gross_revenue         91
recency_days          91
qtde_invoices         91
qtde_items            91
qtde_products         91
frequency             91
qtde_returns           0
avg_ticket            91
avg_recency_days    2816
dtype: int64

## <font color='red'> 4.0. EDA ( Exploratory Data Analysis )</font>

In [38]:
df4 = df_ref.dropna()

### 4.3. Estudo do Espaço

In [57]:
# selected dataset
cols_selected = ['customer_id', 'gross_revenue', 'recency_days', 'qtde_products', 'frequency', 'qtde_returns']
df43 = df4[ cols_selected ].copy()

In [62]:
df43['gross_revenue'].head()

0    0.019292
1    0.011559
2    0.024000
3    0.003375
4    0.003116
Name: gross_revenue, dtype: float64

In [59]:
AWS_ACCESS_KEY_ID = os.environ.get( 'AWS_ACCESS_KEY_ID' )
AWS_SECRET_ACCESS_KEY = os.environ.get( 'AWS_SECRET_ACCESS_KEY' )

In [60]:
print(AWS_ACCESS_KEY_ID)

AKIAW44IREBLQIPGP37Y


In [61]:
mm = pp.MinMaxScaler()

#Creating Session With Boto3.
session = boto3.Session(
aws_access_key_id=AWS_ACCESS_KEY_ID,
aws_secret_access_key=AWS_SECRET_ACCESS_KEY
)

#Creating S3 Resource From the Session.
s3 = session.resource('s3')


df43['gross_revenue'] = mm.fit_transform( df43[['gross_revenue']] )
pickle.dump( mm, open( '../src/features/gross_revenue_scaler.pkl', 'wb') )
result = s3.Bucket('insiders-dataset-tbl').upload_file('../src/features/gross_revenue_scaler.pkl','gross_revenue_scaler.pkl')

df43['recency_days'] = mm.fit_transform( df43[['recency_days']] )
pickle.dump( mm, open( '../src/features/recency_days_scaler.pkl', 'wb') )
result = s3.Bucket('insiders-dataset-tbl').upload_file('../src/features/recency_days_scaler.pkl','recency_days_scaler.pkl')

df43['qtde_products'] = mm.fit_transform( df43[['qtde_products']] )
pickle.dump( mm, open( '../src/features/qtde_products_scaler.pkl', 'wb') )
result = s3.Bucket('insiders-dataset-tbl').upload_file('../src/features/qtde_products_scaler.pkl','qtde_products_scaler.pkl')

df43['frequency'] = mm.fit_transform( df43[['frequency']] )
pickle.dump( mm, open( '../src/features/frequency_scaler.pkl', 'wb') )
result = s3.Bucket('insiders-dataset-tbl').upload_file('../src/features/frequency_scaler.pkl','frequency_scaler.pkl')

df43['qtde_returns'] = mm.fit_transform( df43[['qtde_returns']] )
pickle.dump( mm, open( '../src/features/qtde_returns_scaler.pkl', 'wb') )
result = s3.Bucket('insiders-dataset-tbl').upload_file('../src/features/qtde_returns_scaler.pkl','qtde_returns_scaler.pkl')

### 4.3.3. Tree-Based Embedding

In [49]:
# training dataset
X = df43.drop( columns=['customer_id', 'gross_revenue'], axis=1 )
y = df43['gross_revenue']

# model definition
rf_model = en.RandomForestRegressor( n_estimators=100, random_state=42 )

# model training
rf_model.fit( X, y )
pickle.dump( rf_model, open('../src/models/rf_model.pkl', 'wb'))
result = s3.Bucket('insiders-dataset-tbl').upload_file('../src/models/rf_model.pkl','rf_model.pkl')

# Leaf
df_leaf = pd.DataFrame( rf_model.apply( X ) )

# dataframe Leaf

In [51]:
# Reduzer dimensionality
reducer = umap.UMAP( random_state=42 )
embedding = reducer.fit_transform( df_leaf )

pickle.dump( reducer, open('../src/features/umap_reducer.pkl', 'wb'))
result = s3.Bucket('insiders-dataset-tbl').upload_file('../src/features/umap_reducer.pkl','umap_reducer.pkl')


# embedding
df_tree = pd.DataFrame()
df_tree['embedding_x'] = embedding[:, 0]
df_tree['embedding_y'] = embedding[:, 1]

## <font color='red'> 5.0. Hyperparameter Fine-Tunning </font>

In [52]:
X = df_tree.copy()

In [53]:
X.head()

embedding_x  embedding_y
0    -1.794532    -5.616612
1   -11.619191     3.220745
2    -1.727290    -6.593994
3    12.381402     8.551535
4    15.797197     0.220842

## <font color='red'> 6.0. Model Training </font>

### 6.1. Final Model

In [54]:
# model definition GMM
k = 23
gmm_model = mx.GaussianMixture( n_components=k, n_init=300, random_state = 42 )

# model training
gmm_model.fit( X )

# model predict
labels = gmm_model.predict( X )

### 6.2. Cluster Validation

In [55]:
## SS ( Silhouette Score )
print ( 'SS value: {}'.format( m.silhouette_score( X, labels, metric='euclidean' ) ) )

SS value: 0.6470698118209839


## <font color='red'> 7.0. Cluster Analysis </font>

### 7.1. Cluster Profile

In [60]:
df92 = df4[ cols_selected ].copy()
df92['cluster'] = labels

# change dtypes
df92['recency_days'] = df92['recency_days'].astype( int )
df92['qtde_products'] = df92['qtde_products'].astype( int )
df92['qtde_returns'] = df92['qtde_returns'].astype( int )

In [61]:
# Number of Customer
df_cluster = df92[['customer_id', 'cluster']].groupby('cluster').count().reset_index()
df_cluster['perc_cluster'] = 100 * ( df_cluster['customer_id'] / df_cluster['customer_id'].sum() )

# Average Gross Revenue
df_avg_gross_revenue = df92[['gross_revenue', 'cluster']].groupby('cluster').mean().reset_index()
df_cluster = pd.merge( df_cluster, df_avg_gross_revenue, how='inner', on='cluster')

# Average Recency Days
df_avg_recency_days = df92[['recency_days', 'cluster']].groupby('cluster').mean().reset_index()
df_cluster = pd.merge( df_cluster, df_avg_recency_days, how='inner', on='cluster')

# Average Invoice_no
df_qtde_products = df92[['qtde_products', 'cluster']].groupby('cluster').mean().reset_index()
df_cluster = pd.merge( df_cluster, df_qtde_products, how='inner', on='cluster')

# Frequency
df_frequency = df92[['frequency', 'cluster']].groupby('cluster').mean().reset_index()
df_cluster = pd.merge( df_cluster, df_frequency, how='inner', on='cluster')

# Returns
df_qtde_returns = df92[['qtde_returns', 'cluster']].groupby('cluster').mean().reset_index()
df_cluster = pd.merge( df_cluster, df_qtde_returns, how='inner', on='cluster')

df_cluster.sort_values(by='gross_revenue', ascending=False)

cluster  customer_id  perc_cluster  gross_revenue  recency_days  \
3         3          373     12.567385   10502.792386     19.268097   
5         5          134      4.514825    4595.674627     52.194030   
7         7          145      4.885445    3164.115448     34.427586   
1         1           84      2.830189    2692.810238     76.476190   
11       11          272      9.164420    2315.843676     33.452206   
12       12           95      3.200809    2291.139684     29.957895   
6         6          404     13.611860    1689.073490     54.809406   
0         0          415     13.982480    1221.472892     61.371084   
4         4          279      9.400270     954.046989     77.362007   
8         8          175      5.896226     827.409714     92.022857   
10       10          200      6.738544     647.933500     47.280000   
9         9          226      7.614555     565.176947    139.991150   
2         2          166      5.592992     424.504277    162.271084   

    qtde_products  frequency  qtde_returns  
3      475.528150   0.110615    187.120643  
5      101.134328   0.063555     39.432836  
7      173.841379   0.060415     23.482759  
1      128.511905   0.073652     32.559524  
11     128.316176   0.032634     15.996324  
12     222.115789   0.031576      0.663158  
6       89.175743   0.049992     11.542079  
0       53.219277   0.042040      9.122892  
4       36.530466   0.081823      5.709677  
8       26.811429   0.161952     22.868571  
10      11.510000   0.025991      0.845000  
9       18.681416   0.193812      2.893805  
2       11.198795   0.743519     18.361446

3   Cluster Insiders

5/7 Clusters Quase Insiders

1  Cluster Precisa de Mais Valor "Expend Money"

11   Cluster Mais valor / Frequencia





### Cluster Insiders: ( Cluster 3 )
    - Número de customers: 373 ( 12,5% dos customers )
    - Faturamento médio: 10500
    - Recência em média: 19 dias
    - Média de produtos comprados: 475 produtos
    - Frequência de produtos comprados: 0,11 produtos/dia

### Cluster 02:
    - Número de customers: 31 ( 71% dos customers )
    - Recência em média: 14 dias
    - Compras em média: 53 compras
    - Receita em média: $40.543,52

### Cluster 03:
    - Número de customers: 4335 ( 99% dos customers )
    - Recência em média: 92 dias
    - Compras em média: 5 compras
    - Receita em média: $1.372,57

## <font color='red'> 8.0 Deploy To Production </font>

In [62]:
df92.head()

customer_id  gross_revenue  recency_days  qtde_products  frequency  \
0        17850        5391.21           372            297  17.000000   
1        13047        3232.59            56            171   0.028302   
2        12583        6705.38             2            232   0.040323   
3        13748         948.25            95             28   0.017921   
4        15100         876.00           333              3   0.073171   

   qtde_returns  cluster  
0            40        3  
1            35        7  
2            50        3  
3             0        8  
4            22        2

In [63]:
df92.dtypes

customer_id        int64
gross_revenue    float64
recency_days       int64
qtde_products      int64
frequency        float64
qtde_returns       int64
cluster            int64
dtype: object

### 8.1 Inserto into SQLITE

In [90]:
import sqlite3
from sqlalchemy import create_engine

In [91]:
df92['recency_days'] = df92['recency_days'].astype( int )
df92['qtde_products'] = df92['qtde_products'].astype( int )
df92['qtde_returns'] = df92['qtde_returns'].astype( int )

#df92['last_training_timestamp'] = datetime.now().strftime( '%Y-%m-%d %H:%M%S' )

In [93]:
# create table
query_create_table_insiders = """
    CREATE TABLE insiders (
        customer_id     INTEGER,
        gross_revenue   REAL,
        recency_days    INTEGER,
        qtde_products   INTEGER,
        frequency       REAL,
        qtde_returns    INTEGER,
        cluster         INTEGER
    )
"""    
conn = sqlite3.connect( 'insiders_db.sqlite')
conn.execute( query_create_table_insiders )
conn.commit()
conn.close()

# insert data
conn = create_engine( 'sqlite:///insiders_db.sqlite')
df92.to_sql( 'insiders', con=conn, if_exists='append', index=False )



# select data

2968

In [94]:
# consulting database
query = """
    SELECT * FROM insiders
"""

df = pd.read_sql_query( query, conn )

In [95]:
df.head()

customer_id  gross_revenue  recency_days  qtde_products  frequency  \
0        17850        5391.21           372            297  17.000000   
1        13047        3232.59            56            171   0.028302   
2        12583        6705.38             2            232   0.040323   
3        13748         948.25            95             28   0.017921   
4        15100         876.00           333              3   0.073171   

   qtde_returns  cluster  
0            40        3  
1            35        7  
2            50        3  
3             0        8  
4            22        2

In [96]:
df.shape

(2968, 7)